server/src/test/java/org/elasticsearch/action/support/TransportActionFilterChainRefCountingTests.java

```diff
@@ -0,0 +1,209 @@
+        final var request = new Request();
+        try {
+            client().execute(TYPE, request, ActionListener.<Response>running(countDownLatch::countDown).delegateResponse((delegate, e) -> {
+                // _If_ we got an exception then it must be an ElasticsearchException with message "short-circuit failure", i.e. we're
+                // checking that nothing else can go wrong here. But it's also ok for everything to succeed too, in which case we countDown
+                // the latch without running this block.
+                assertEquals("short-circuit failure", asInstanceOf(ElasticsearchException.class, e).getMessage());
+                delegate.onResponse(null);
+            }));
+        } finally {
+            request.decRef();
+        }
+        request.addCloseListener(ActionListener.running(countDownLatch::countDown));
+        safeAwait(countDownLatch);
+    }
+
+    public static class TestPlugin extends Plugin implements ActionPlugin {
+
+        private ThreadPool threadPool;
+
+        @Override
+        public Collection<?> createComponents(PluginServices services) {
+            threadPool = services.threadPool();
+            return List.of();
+        }
+
+        @Override
+        public List<ActionHandler<? extends ActionRequest, ? extends ActionResponse>> getActions() {
+            return List.of(new ActionHandler<>(TYPE, TestAction.class));
+        }
+
+        @Override
+        public List<ActionFilter> getActionFilters() {
+            return randomSubsetOf(
+                List.of(
+                    new TestAsyncActionFilter(threadPool),
+                    new TestAsyncActionFilter(threadPool),
+                    new TestAsyncMappedActionFilter(threadPool),
+                    new TestAsyncMappedActionFilter(threadPool)
+                )
+            );
+        }
+    }
+
+    private static class TestAsyncActionFilter implements ActionFilter {
+
+        private final ThreadPool threadPool;
+        private final int order = randomInt();
+
+        private TestAsyncActionFilter(ThreadPool threadPool) {
+            this.threadPool = Objects.requireNonNull(threadPool);
+        }
+
+        @Override
+        public int order() {
+            return order;
+        }
+
+        @Override
+        public <Req extends ActionRequest, Rsp extends ActionResponse> void apply(
+            Task task,
+            String action,
+            Req request,
+            ActionListener<Rsp> listener,
+            ActionFilterChain<Req, Rsp> chain
+        ) {
+            if (action.equals(TYPE.name())) {
+                randomFrom(EsExecutors.DIRECT_EXECUTOR_SERVICE, threadPool.generic()).execute(new AbstractRunnable() {
+                    @Override
+                    public void onFailure(Exception e) {
+                        fail(e);
+                    }
+
+                    @Override
+                    protected void doRun() {
+                        Thread.yield();
```

Comment:
Is this meant to mimic work?

Reply:
Kinda, but really I was just using it to make sure I had the right things covered by the tests and forgot to remove it. Gone now.